service/header/sync.go

```diff
@@ -20,147 +23,263 @@ type Syncer struct {
+		return pubsub.ValidationIgnore // we don't know if header is invalid so ignore
+	}
 
-// getHead tries to get head locally and if not exists requests trusted hash.
-func (s *Syncer) getHead(ctx context.Context) (*ExtendedHeader, error) {
-	head, err := s.store.Head(ctx)
-	switch err {
-	case nil:
-		return head, nil
-	case ErrNoHead:
-		// if there is no head - request header at trusted hash.
-		trusted, err := s.exchange.RequestByHash(ctx, s.trusted)
-		if err != nil {
-			log.Errorw("requesting header at trusted hash", "err", err)
-			return nil, err
-		}
+	// 3. Filter out maybeHead if behind trusted
+	if maybeHead.Height <= trstHead.Height {
+		log.Warnw("received known header",
+			"height", maybeHead.Height,
+			"hash", maybeHead.Hash(),
+			"from", p.ShortString())
 
-		err = s.store.Append(ctx, trusted)
-		if err != nil {
-			log.Errorw("appending header at trusted hash to store", "err", err)
-			return nil, err
-		}
+		// TODO(@Wondertan): Remove once duplicates are fully fixed
+		log.Warnf("Ignore the warn above - there is a known issue with duplicate headers on the network.")
+		return pubsub.ValidationIgnore // we don't know if header is invalid so ignore
+	}
 
-		return trusted, nil
+	// 4. Verify maybeHead against trusted
+	err = trstHead.VerifyNonAdjacent(maybeHead)
+	var verErr *VerifyError
+	if errors.As(err, &verErr) {
+		log.Errorw("invalid header",
+			"height", maybeHead.Height,
+			"hash", maybeHead.Hash(),
+			"from", p.ShortString(),
+			"reason", verErr.Reason)
+		return pubsub.ValidationReject
 	}
 
-	return nil, err
+	// 5. Save verified header to pending cache
+	// NOTE: Pending cache can't be DOSed as we verify above each header against a trusted one.
+	s.pending.Add(maybeHead)
+	// and trigger sync to catch-up
+	s.wantSync()
+	log.Infow("new pending head",
+		"height", maybeHead.Height,
+		"hash", maybeHead.Hash(),
+		"from", p.ShortString())
+	return pubsub.ValidationAccept
 }
 
 // TODO(@Wondertan): Number of headers that can be requested at once. Either make this configurable or,
-// find a proper rationale for constant.
-var requestSize uint64 = 128
+//  find a proper rationale for constant.
+var requestSize uint64 = 512
 
-// syncDiff requests headers from knownHead up to new head.
-func (s *Syncer) syncDiff(ctx context.Context, knownHead, newHead *ExtendedHeader) error {
-	start, end := uint64(knownHead.Height+1), uint64(newHead.Height)
-	for start < end {
-		amount := end - start
+// syncTo requests headers from locally stored head up to the new head.
+func (s *Syncer) syncTo(ctx context.Context, newHead *ExtendedHeader) error {
+	head, err := s.store.Head(ctx)
+	if err != nil {
+		return err
+	}
+
+	if head.Height == newHead.Height {
+		return nil
+	}
+
+	log.Infow("syncing headers", "from", head.Height, "to", newHead.Height)
+	defer log.Info("synced headers")
+
+	start, end := uint64(head.Height)+1, uint64(newHead.Height)
+	for start <= end {
+		amount := end - start + 1
```

Comment:
Ah, I see now. This is where amount could end up being 0, right?

Reply:
It can be 0 on line 288